### ISSUE TITLE
On the usage of curl

### ISSUE THREAD

**Issue by nagy** [minad/osm#20]
I have a question. Why was curl used and not the builtin `url-retrieve-synchronously`? I assume it was because it would block emacs and cause a huge lag without any display updates? There is also `url-retrieve` which seems to be the asynchronous alternative to that. Also there is the `request` library, which may handle things a bit better and with less code than the builtin ones. https://github.com/tkf/emacs-request

If you want to keep using curl, there are however some improvements to make. I am seeing that you are starting one curl command for each tile. This negotiates a new TLS handshake for each tile with the server and therefore taxes the server a little bit.

Curl supports to query multiple urls at once. For this we would need to know all the urls that we need to fetch in advance, but I think we do. You can then do something like:

```
curl --fail --silent --output outputfile1.png https://.../url1.png --output outputfile2 https://.../url2.png
```

Then, curl has the chance to reuse the same connection for all urls. Be aware, that some options are global and some need to be repeated for all urls, like `--output`.

If you agree to all of this, then there is another potential for improvement. Curl is also able to fetch all urls in parallel using the `--parallel` option.

https://curl.se/docs/manpage.html

I think the combination of this has the potential to significantly increase the speed with which tiles are downloaded and displayed.

**Comment by minad** [comment]
1. `url-retrieve-synchrously` cannot be used due to the blocking nature
2. `url-retrieve` could be used. On the other hand `curl` removes pressure from Emacs itself.
3. The request package would be an option too but it doesn't offer big benefits as far as I can tell. I try to avoid dependencies.
4. `curl --parallel` seems to be a new addition to curl. I think it is not available in Debian stable. Therefore we cannot use this. Also it won't help with reducing the overhead since parallel connections need a separate handshake anyway and we also want to connect to multiple servers in around robin fashion.
5. `curl`ing multiple urls sequentially is a possibility. It may make sense to do this to reduce the TLS handshake overhead. Maybe one could use one curl instance to fetch 4 or 8 tiles (or a configurable number) from one server sequentially. This should indeed reduce the load on the server and improve responsiveness. Feel free to give this a try in a PR. It needs some care to get this right, for example one has to fetch tile 1, 7, 13 and 19 as one batch to get the queue order right if we have 6 parallel connections (3 servers, 2 connections). Furthermore the tiles should be updated on the display as soon as they come in which complicates the process handling. Currently I only use a simple sentinel.

**Comment by nagy** [comment]
Thank you for the explanation.


> The request package would be an option too but it doesn't offer big benefits as far as I can tell. I try to avoid dependencies.

It would drop the dependency on curl, but since curl is pretty wide spread I dont think this would be worth the effort

> curl --parallel seems to be a new addition to curl. I think it is not available in Debian stable. Therefore we cannot use this.

This could of course be only optionally enabled if the underlying curl supports it.

> Also it won't help with reducing the overhead since parallel connections need a separate handshake anyway.

No, with HTTP/2 multiple requests can be made within the same connection with what they call "streams". 

> and we also want to connect to multiple servers in a round robin fashion.

We can still do the round robin on a time basis or on each full queue iteration.

> Furthermore the tiles should be updated on the display as soon as they come in

This of course would not be possible anymore. Overall they come in quicker but it may not seem as responsible. It is certainly a trade-off.

**Comment by minad** [comment]
> It would drop the dependency on curl, but since curl is pretty wide spread I dont think this would be worth the effort

request.el wouldn't give us benefits since it still relies on curl or url.el. Nothing would be won.

> No, with HTTP/2 multiple requests can be made within the same connection with what they call "streams".

I am not familiar with HTTP/2 but what you write with regards to streams sounds plausible. Note however that three handshakes will always be needed for the servers a, b and c. Given that this load balancing design already forces us to connect to multiple servers at once I wonder if the osm tile servers have been designed with streams in mind. Do they usually offer HTTP/2?

I would be okay if we use the parallel curl feature if available.

> We can still do the round robin on a time basis or on each full queue iteration.

Of course, we would keep that.

> This of course would not be possible anymore. Overall they come in quicker but it may not seem as responsible. It is certainly a trade-off.

Well, it is possible. We would have to parse the curl output or check the filesystem via the notify api. I definitely wouldn't want to lose the immediate tile display.

**Comment by nagy** [comment]
> since it still relies on curl 

oh, I did not know that.

> Note however that three handshakes will always be needed

Maybe my previous explanation was not clear enough. Lets say the user moves a bit to the right and now 10 more tiles need to be fetched. Then we fetch all these 10 tiles with one curl command ( and optionally `--parallel` ) from one of the tiles servers, lets say "a". Then the curl command finishes and the TCP connection is closed. Then some time passes and the user moves again. Now we need to fetch another 10 tiles but this time, we choose server "b". 

This should be a fair approach to balance out the load on the servers and the amount handshakes is the same as if they would only offer one server. So we dont need to do multiple concurrent handshakes.

> Do they usually offer HTTP/2

A quick test shows that they offer HTTP/2. Of course they could still only sequentially server the streams and I have not tested that but that would surprise me. But even if they did, they may only do so now and the clients just normally fetch via HTTP/2 that would give them the chance to fix this issue on their side later on.

> Using keep-alive ...

The keep-alive is not of much use if the curl process exits immediately.

> I definitely don't want to lose the immediate tile display.

Okay, I hope the notify api only notifies us if the png file is written out completely.

**Comment by nagy** [comment]
Btw, fetching multiple urls via curl and running an update code on each file when the filesystem notification api tells us that the file is complete....

I cannot recognize anything openstreetmap specific in this, so it may not even be a good place to implement this here, but rather factor it out in an external package. But that is of course another dependency.

**Comment by minad** [comment]
> Maybe my previous explanation was not clear enough. Lets say the user
moves a bit to the right and now 10 more tiles need to be fetched. Then
we fetch all these 10 tiles with one curl command ( and optionally
`--parallel` ) from one of the tiles servers, lets say "a". Then the
curl command finishes and the TCP connection is closed. Then some time
passes and the user moves again. Now we need to fetch another 10 tiles
but this time, we choose server "b".
>
> This should be a fair approach to balance out the load on the servers
and the amount handshakes is the same as if they would only offer one
server. So we dont need to do multiple concurrent handshakes.

Okay, I wouldn't do it like this. As far as I understood, according to
the osm server guidelines we are allowed to access the servers a, b and
c in parallel. If we want to fetch 12 tiles, we fetch 4 from a
sequentially, 4 from b sequentially, 4 from c sequentially. This should
give the best result.

>> Do they usually offer HTTP/2
>
> A quick test shows that they offer HTTP/2. Of course they could still
only sequentially server the streams and I have not tested that but that
would surprise me. But even if they did, they may only do so now and the
clients just normally fetch via HTTP/2 that would give them the chance
to fix this issue on their side later on.

Of course. I agree with that reasoning. But I still suspect that we will
get diminishing returns. Also the server load won't be reduced
significantly, given that the server still has to handle two parallel
stream connections.

>> Using keep-alive ...
>
> The keep-alive is not of much use if the curl process exits immediately.

Of course. My statement applies to using a single curl process to fetch
a batch of tiles sequentially as described above. Then curl will keep
the connection alive, avoiding the expensive handshake.

>> I definitely don't want to lose the immediate tile display.
>
> Okay, I hope the notify api only notifies us if the png file is
written out completely.
I don't think that's the case. I would probably go with parsing the curl
output, which should be more robust. Try this:

curl --silent --write-out '%{http_code} %{filename_effective}\n' -o
file1 url1 -o file2 url2

**Comment by minad** [comment]
> Maybe my previous explanation was not clear enough. Lets say the user moves a bit to the right and now 10 more tiles need to be fetched. Then we fetch all these 10 tiles with one curl command ( and optionally `--parallel` ) from one of the tiles servers, lets say "a". Then the curl command finishes and the TCP connection is closed. Then some time passes and the user moves again. Now we need to fetch another 10 tiles but this time, we choose server "b". 
> This should be a fair approach to balance out the load on the servers and the amount handshakes is the same as if they would only offer one server. So we dont need to do multiple concurrent handshakes.

Okay, I wouldn't do it like this. As far as I understood, according to
the osm server guidelines we are allowed to access the servers a, b and
c in parallel. If we want to fetch 12 tiles, we fetch 4 from a
sequentially, 4 from b sequentially, 4 from c sequentially. This should
give the best result.

> A quick test shows that they offer HTTP/2. Of course they could still only sequentially server the streams and I have not tested that but that would surprise me. But even if they did, they may only do so now and the clients just normally fetch via HTTP/2 that would give them the chance to fix this issue on their side later on.

Of course. I agree with that reasoning. But I still suspect that we will
get diminishing returns. Also the server load won't be reduced
significantly, given that the server still has to handle two parallel
stream connections.

> The keep-alive is not of much use if the curl process exits immediately.

Of course. My statement applies to using a single curl process to fetch
a batch of tiles sequentially as described above. Then curl will keep
the connection alive, avoiding the expensive handshake.

> Okay, I hope the notify api only notifies us if the png file is written out completely. 

I don't think that's the case. I would probably go with parsing the curl
output, which should be more robust. Try this:

curl --silent --write-out '%{http_code} %{filename_effective}\n' -o
file1 url1 -o file2 url2

**Comment by minad** [comment]
> Btw, fetching multiple urls via curl and running an update code on each file when the filesystem notification api tells us that the file is complete....

See what I wrote in my other comment:

curl --silent --write-out '%{http_code} %{filename_effective}\n' -o
file1 url1 -o file2 url2

I would avoid the notify API here and instead parse the process output
which we can control via `--write-out`.

> I cannot recognize anything openstreetmap specific in this, so it may not even be a good place to implement this here, but rather factor it out in an external package. But that is of course another dependency.

Yes, there is nothing OSM-specific here besides our requirements with
regards to scheduling. It would be nice to have a separate URL
fetching/scheduling package with all the features we are discussing here
(parallel and sequential fetching, notification of individual download
status per file, etc.).

However I am not really interested in writing such a base library. It
may make sense to offer this to Emacs itself for inclusion, which means
a lot of discussion and a lot of work. So I would prefer to not do that
and focus on osm.el for now. One could still take the code out from
osm.el later on and make it a freestanding library.

**Comment by minad** [comment]
> I cannot recognize anything openstreetmap specific in this, so it may not even be a good place to implement this here, but rather factor it out in an external package. But that is of course another dependency.

Note that we are not the first here to reinvent fetching via curl.

* https://github.com/tkf/emacs-request
* https://github.com/skeeto/elfeed/blob/master/elfeed-curl.el
* https://github.com/alphapapa/plz.el

I am sure there are even more. It would be nice to have a general
library which does it all and which does it right, but this is not an
easy task, even if one doesn't aim to establish such a library as the
standard url fetching package (on ELPA or Emacs core). Or one doesn't
aim to write an all-encompassing (over-engineered) library, staying more
modest, solving only the one use case we have here. But then what is the
point? Do we have a second use case in sight were we need this exact
feature? I don't have such another project planned and if it comes up I
could still extract and share the relevant code in a separate library.

**Comment by minad** [comment]
I added a basic implementation which fetches the tiles in batches such that the https connection is kept alive for longer. This speeds up loading quite a bit. We could still explore further improvements, e.g., using --parallel or improving the scheduling.

**Comment by nagy** [comment]
Wow, thank you. This does already go a large step into the right direction.

Yes, a dedicated library would be a lot of effort. Too much for this small scope. Maybe emacs will have something like this one day.

**Comment by nagy** [comment]
I just tested it out. It now loads incredibly faster than before. Even faster than in my webbrowser. Great work!
